Signal/src/view controllers/RegistrationViewController.m

```diff
@@ -148,48 +148,60 @@ - (void)dismissKeyboardFromAppropriateSubView {
 
 - (BOOL)textField:(UITextField *)textField
     shouldChangeCharactersInRange:(NSRange)range
-                replacementString:(NSString *)string {
-    NSString *textBeforeChange = textField.text;
-
-    // backspacing should skip over formatting characters
-    UITextPosition *posIfBackspace = [textField positionFromPosition:textField.beginningOfDocument
-                                                              offset:(NSInteger)(range.location + range.length)];
-    UITextRange *rangeIfBackspace = [textField textRangeFromPosition:posIfBackspace toPosition:posIfBackspace];
-    bool isBackspace =
-        string.length == 0 && range.length == 1 && [rangeIfBackspace isEqual:textField.selectedTextRange];
-    if (isBackspace) {
-        NSString *digits                       = textBeforeChange.digitsOnly;
-        NSUInteger correspondingDeletePosition = [PhoneNumberUtil translateCursorPosition:range.location + range.length
-                                                                                     from:textBeforeChange
-                                                                                       to:digits
-                                                                        stickingRightward:true];
-        if (correspondingDeletePosition > 0) {
-            textBeforeChange = digits;
-            range            = NSMakeRange(correspondingDeletePosition - 1, 1);
-        }
-    }
-
-    // make the proposed change
-    NSString *textAfterChange            = [textBeforeChange withCharactersInRange:range replacedBy:string];
-    NSUInteger cursorPositionAfterChange = range.location + string.length;
-
+                replacementString:(NSString *)insertionText {
+    
+    // Phone numbers takes many forms.
+    //
+    // * We only want to let the user enter decimal digits.
+    // * The user shouldn't have to enter hyphen, parentheses or whitespace;
+    //   the phone number should be formatted automatically.
+    // * The user should be able to copy and paste freely.
+    // * Invalid input should be simply ignored.
+    //
+    // We accomplish this by being permissive and trying to "take as much of the user
+    // input as possible".
+    //
+    // * Always accept deletes.
+    // * Ignore invalid input.
+    // * Take partial input if possible.
+    
+    NSString *oldText = textField.text;
```

Comment:
nit: Are the changes to this method fixing any known problem? Even though your implementation is arguably cleaner, it seems like code that would be easy to get wrong. If it ain't broke, yadda yadda.